lib/std/os.zig

```diff
@@ -6281,44 +6283,68 @@ var has_copy_file_range_syscall = std.atomic.Atomic(bool).init(true);
+    // WASI-libc doesn't have copy_file_range.
+    if (builtin.os.tag != .wasi) {
+        switch (builtin.os.tag) {
+            .linux => {
+                if (std.c.versionCheck(.{ .major = 2, .minor = 27, .patch = 0 }).ok and
+                    builtin.os.isAtLeast(.linux, .{ .major = 4, .minor = 5 }) orelse true and
+                    has_copy_file_range_syscall.load(.Monotonic))
+                {
+                    var off_in_copy = @bitCast(i64, off_in);
+                    var off_out_copy = @bitCast(i64, off_out);
+
+                    const rc = system.copy_file_range(fd_in, &off_in_copy, fd_out, &off_out_copy, len, flags);
+                    switch (system.getErrno(rc)) {
+                        .SUCCESS => return @intCast(usize, rc),
+                        .BADF => return error.FilesOpenedWithWrongFlags,
+                        .FBIG => return error.FileTooBig,
+                        .IO => return error.InputOutput,
+                        .ISDIR => return error.IsDir,
+                        .NOMEM => return error.OutOfMemory,
+                        .NOSPC => return error.NoSpaceLeft,
+                        .OVERFLOW => return error.Unseekable,
+                        .PERM => return error.PermissionDenied,
+                        .TXTBSY => return error.SwapFile,
+                        // these may not be regular files, try fallback
+                        .INVAL => {},
+                        // support for cross-filesystem copy added in Linux 5.3, use fallback
+                        .XDEV => {},
+                        // syscall added in Linux 4.5, use fallback
+                        .NOSYS => {
+                            has_copy_file_range_syscall.store(false, .Monotonic);
+                        },
+                        else => |err| return unexpectedErrno(err),
+                    }
+                }
             },
-            else => |err| return unexpectedErrno(err),
+            .freebsd => {
+                if (builtin.os.isAtLeast(.freebsd, .{ .major = 13, .minor = 0 }) orelse true) {
+                    var off_in_copy = @bitCast(i64, off_in);
+                    var off_out_copy = @bitCast(i64, off_out);
+                    const rc = freebsd.copy_file_range(fd_in, &off_in_copy, fd_out, &off_out_copy, len, flags);
+                    switch (system.getErrno(rc)) {
+                        .SUCCESS => return @intCast(usize, rc),
+                        .BADF => return error.FilesOpenedWithWrongFlags,
+                        .FBIG => return error.FileTooBig,
+                        .INVAL => return error.RangeOverlap,
+                        .IO => return error.InputOutput,
+                        .INTEGRITY => return error.CorruptedData,
+                        .ISDIR => return error.IsDir,
+                        .NOSPC => return error.NoSpaceLeft,
+                        .INTR => unreachable,
```

Comment:
Many of the other functions in `os` will wrap functions in a infinite `while` loop that will `continue` when `EINTR` is returned. It would probably be better to do that here to preserve the same semantics as on Linux.